### PR TITLE
Options revamp

### PR DESCRIPTION
**Pull request by wthrowe**
* Simplifies the interfaces for declaring constructible objects.
* Decreases the number of places the entire option parser needs to be included to reduce compile time a bit.
* Adds an option parsing tutorial.

---
This change is [Reviewable](https://reviewable.io/reviews/sxs-collaboration/spectre/322)

### CODE REVIEW

File: src/Options/ParseOptions.hpp [hunk 1]
```diff
@@ -0,0 +1,606 @@
...
+#include "Utilities/Overloader.hpp"
+#include "Utilities/PrettyType.hpp"
+#include "Utilities/Requires.hpp"
+#include "Utilities/StdHelpers.hpp"
+#include "Utilities/TypeTraits.hpp"
+
+/// \ingroup OptionParsing
+/// The type that options are passed around as.  Contains YAML node
+/// data and an OptionContext.
+class Option_t {
```
Comment (nilsdeppe) [inline src/Options/ParseOptions.hpp]: The `_t` should be only used for denoting eager metafunctions, not classes

File: tests/Unit/TestFactoryCreation.hpp [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+namespace TestFactoryCreation_detail {
+template <typename T>
+struct Opt {
+  using type = std::unique_ptr<T>;
+  static constexpr OptionString_t help = {"halp"};
+};
+}  // namespace TestFactoryCreation_detail
+
+/// Construct a factory object from a given string.  Each line in the
```
Comment (nilsdeppe) [inline tests/Unit/TestFactoryCreation.hpp]: `\ingroup TestingFramework`

File: docs/Tutorials/Tutorials.md [hunk 3]
```diff
@@ -6,3 +6,4 @@ See LICENSE.txt for details.
 
 - \ref visualization - How to visualize volume data
 - \ref tuts_orientations - How to determine orientations
+- \ref tuts_option_parsing - How to get options from input files
```
Comment (kidder) [inline docs/Tutorials/Tutorials.md]: let's spell out tutorial

File: docs/Tutorials/OptionParsing.md [hunk 4]
```diff
@@ -59,13 +60,8 @@ derived classes that can be created.
...
  using creatable_classes = tmpl::list<Derived1, ...>;
 \endcode
-
-Derived classes should:
--# define `static OptionString_t help` containing class-specific help
-   text
--# define a type `options` as a typelist of option structs required to
-   create the class
--# define a constructor taking those options and an OptionContext
+The requested derived class is created in the same was as an
```
Comment (kidder) [inline docs/Tutorials/OptionParsing.md]: was -> way

**Review by kidder — CHANGES_REQUESTED** [review]

File: src/Options/ParseOptions.hpp (hunk 1 above)
Comment (kidder) [reply src/Options/ParseOptions.hpp]: Option_t -> Option

File: docs/Tutorials/OptionParsing.md [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+is all included in the generated help output.
+
+Examples:
+\snippet Test_Options.cpp options_example_scalar_struct
+\snippet Test_Options.cpp options_example_vector_struct
+
+The option type can be any type understood natively by yaml-cpp
+(fundamentals and `std::vector`, `std::list`, and `std::map` of
+parsable types) and types SpECTRE adds support for.  SpECTRE adds
+`std::unordered_map` (but only of ordered types), and various classes
```
Comment (kidder) [inline docs/Tutorials/OptionParsing.md]: only with ordered keys

File: docs/Tutorials/OptionParsing.md [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+marked as constructible in their declarations.
+
+## Constructible classes
+
+A class can be marked as constructible using the
+MAKE_CREATABLE_FROM_YAML() macro (defined in
+`Options/MakeCreatableFromYaml.hpp`.  The creatable class must define
+`static OptionString_t help` and a typelist of option structs
```
Comment (kidder) [inline docs/Tutorials/OptionParsing.md]: constexpr

File: docs/Tutorials/OptionParsing.md [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+Unlike option descriptions, which should be brief, the class help
+string has no length limits and should give a description of the class
+and any necessary discussion of its options beyond what can be
+described in their individual help strings.
+
+Creatable classes must be default constructible and move assignable.
+
+Example:
+\snippet Test_MakeCreatableFromYaml.cpp MCFY_example
```
Comment (kidder) [inline docs/Tutorials/OptionParsing.md]: add comment to snippet making clear that CFO would be used in the input file to create a CreateFromOptions

File: docs/Tutorials/OptionParsing.md [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+
+## Factory
```
Comment (kidder) [inline docs/Tutorials/OptionParsing.md]: there is a was that should be a way

**Review by kidder — CHANGES_REQUESTED** [review]

**Review by kidder — COMMENTED** [review]
I am happy with the latest changes

**Review by kidder — APPROVED** [review]

**Review by nilsdeppe — APPROVED** [review]

**Pull request merged**